Python/Product/Uwp.Interpreter/PythonUwpInterpreterFactoryProvider.cs

```diff
@@ -21,106 +21,381 @@
         }
 
-        private void DiscoverFactories() {
-            if (_factories == null) {
-                _factories = new HashSet<IPythonInterpreterFactory>();
+        /// <summary>
+        /// Gets information about an "in-memory" project.  Supports reading interpreters from
+        /// a project when we're out of proc that haven't yet been committed to disk.
+        /// </summary>
+        private class InMemoryProjectInfo : ProjectInfo
+        {
+            private readonly InMemoryProject _project;
+
+            public InMemoryProjectInfo(InMemoryProject project, IProjectContextProvider contextProvider) 
+                : base(contextProvider) {
+                _project = project;
+            }
 
-                var userSdkInstallDir = Environment.ExpandEnvironmentVariables(@"%LOCALAPPDATA%\Microsoft\VisualStudio\%VISUALSTUDIOVERSION%Exp\Extensions\Microsoft\Python UWP");
-                var sdkInstallDir = Environment.ExpandEnvironmentVariables(@"%ProgramFiles(x86)%\Microsoft SDKs\Windows\v10.0\ExtensionSDKs\Python UWP");
+            public override object Context {
+                get {
+                    return _project;
+                }
+            }
 
-                var factoryMap = FindFactoriesFromDirectories(sdkInstallDir, userSdkInstallDir);
+            public override string FullPath {
+                get {
+                    return _project.FullPath;
+                }
+            }
 
-                if (factoryMap.Count > 0) {
-                    foreach (var factory in factoryMap.Values) {
-                        _factories.Add(factory);
+            public override string GetPropertyValue(string name)
+            {
+                object res;
+                if (_project.Properties.TryGetValue(name, out res) && res is string) {
+                    return (string)res;
+                }
+
+                return String.Empty;
+            }
+        }
+
+        /// <summary>
+        /// Tracks data about a project.  Specific subclasses deal with how the underlying project
+        /// is being stored.
+        /// </summary>
+        private abstract class ProjectInfo : IDisposable {
+            public readonly IProjectContextProvider ContextProvider;
+            private IPythonInterpreterFactory _factory;
+            public IPythonInterpreterFactory Factory {
+                get {
+                    return _factory;
+                }
+                private set {
+                    if (_factory != value) {
+                        var disp = _factory as IDisposable;
+                        if (disp != null) {
+                            disp.Dispose();
+                        }
+                        _factory = value;
                     }
+                }
+            }
+
+            static public ProjectInfo CreateFromProjectContext(object context, IProjectContextProvider contextProvider) {
+                var projContext = context as MSBuild.Project;
```

Comment:
Also see [this change](https://github.com/Microsoft/PTVS/commit/8abb345b055611af9a6cbef52372548b5c4eb708), where I had to fix crashes due to the MSBuild assemblies not being loadable.